### PR TITLE
[CALCITE-4223] Introducing column statistics

### PR DESCRIPTION
**Pull request by chunweilei**
JIRA: https://issues.apache.org/jira/browse/CALCITE-4223

### CODE REVIEW

File: core/src/main/java/org/apache/calcite/plan/RelOptTable.java [hunk 1]
```diff
@@ -133,6 +134,16 @@
    *  same number of entries as there are fields, and is immutable. */
   List<ColumnStrategy> getColumnStrategies();
 
+  /**
+   * Returns column statistics.
+   *
+   * @param colName column names
```
Comment (liyafan82) [inline core/src/main/java/org/apache/calcite/plan/RelOptTable.java:140]: nit: names -> name

File: core/src/main/java/org/apache/calcite/rel/metadata/RelMdDistinctRowCount.java [hunk 2]
```diff
@@ -79,6 +82,37 @@ public Double getDistinctRowCount(RelNode rel, RelMetadataQuery mq,
     return null;
   }
 
+  public Double getDistinctRowCount(TableScan rel, RelMetadataQuery mq,
+      ImmutableBitSet groupKey, RexNode predicate) {
+    final double selectivity = predicate == null ? 1D : mq.getSelectivity(rel, predicate);
+    final boolean uniq = RelMdUtil.areColumnsDefinitelyUnique(mq, rel, groupKey);
+    if (uniq) {
+      return NumberUtil.multiply(mq.getRowCount(rel), selectivity);
```
Comment (liyafan82) [inline core/src/main/java/org/apache/calcite/rel/metadata/RelMdDistinctRowCount.java:90]: If the column values are unique, shouldn't we return 1 * selectivity?

File: core/src/main/java/org/apache/calcite/rel/metadata/RelMdDistinctRowCount.java [hunk 3]
```diff
@@ -79,6 +82,37 @@ public Double getDistinctRowCount(RelNode rel, RelMetadataQuery mq,
...
+
+        final Long cnt = statistics.getCountDistinct();
+        if (cnt == null) {
+          return null;
+        } else {
+          ndv *= cnt;
+        }
+      }
+      final Double rowCount = mq.getRowCount(rel);
+      return (rowCount == null ? ndv : Math.min(ndv, mq.getRowCount(rel))) * selectivity;
```
Comment (liyafan82) [inline core/src/main/java/org/apache/calcite/rel/metadata/RelMdDistinctRowCount.java:112]: The product of individual column NDVs seems an overestimate of the multiple column NDV. We should adjust it with the `RelMdUtil.numDistinctVals` method?

File: core/src/main/java/org/apache/calcite/rel/metadata/RelMdDistinctRowCount.java (hunk 2 above)
Comment (chunweilei) [reply core/src/main/java/org/apache/calcite/rel/metadata/RelMdDistinctRowCount.java:90]: No. If the column values are unique, it does not mean only one row.

File: core/src/main/java/org/apache/calcite/schema/ColStatistics.java [hunk 4]
```diff
@@ -0,0 +1,40 @@
...
+ * Column statistics about a {@link Table}.
+ *
+ * <p>Each of the methods may return {@code null} meaning "not known".</p>
+ */
+public interface ColStatistics extends Wrapper {
+
+  /**
+   * Number of distinct values.
+   */
+  Long getCountDistinct();
```
Comment (vlsi) [inline core/src/main/java/org/apache/calcite/schema/ColStatistics.java:29]: Should all the methods have default implementation like `return null`?

File: core/src/main/java/org/apache/calcite/schema/Statistics.java [hunk 5]
```diff
@@ -82,12 +84,25 @@ public static Statistic of(final double rowCount,
...
+    return of(rowCount, ImmutableList.of(), ImmutableList.of(), ImmutableList.of(), colStatistics);
+  }
+
   /** Returns a statistic with a given row count, set of unique keys,
    * referential constraints, and collations. */
   public static Statistic of(final Double rowCount,
       final List<ImmutableBitSet> keys,
       final List<RelReferentialConstraint> referentialConstraints,
       final List<RelCollation> collations) {
+    return of(rowCount, keys, referentialConstraints, collations, new HashMap<>());
```
Comment (vlsi) [inline core/src/main/java/org/apache/calcite/schema/Statistics.java:98]: `ImmutableMap.of()`?

File: core/src/test/java/org/apache/calcite/test/RelMetadataTest.java [hunk 6]
```diff
@@ -3102,6 +3108,69 @@ private void checkNodeTypeCount(String sql, Map<Class<? extends RelNode>, Intege
...
+      @Override public Double getAvgColLen() {
+        return 2D;
+      }
+
+      @Override public Long getNumNulls() {
+        return null;
+      }
+
+      @Override public <C> C unwrap(Class<C> aClass) {
+        return null;
```
Comment (vlsi) [inline core/src/test/java/org/apache/calcite/test/RelMetadataTest.java:3138]: I believe `unwrap` should have a default implementation.

File: core/src/test/java/org/apache/calcite/test/RelMetadataTest.java [hunk 7]
```diff
@@ -3102,6 +3108,69 @@ private void checkNodeTypeCount(String sql, Map<Class<? extends RelNode>, Intege
...
+      }
+    };
+
+    final RelNode rel = LogicalTableScan.create(cluster, t1, ImmutableList.of());
+    final RelMetadataQuery mq = rel.getCluster().getMetadataQuery();
+
+    ImmutableBitSet groupKey =
+        ImmutableBitSet.of(rel.getRowType().getFieldNames().indexOf(col1));
+    Double ndv = mq.getDistinctRowCount(rel, groupKey, null);
+    assertThat(ndv.intValue(), is(5));
```
Comment (vlsi) [inline core/src/test/java/org/apache/calcite/test/RelMetadataTest.java:3168]: Please add the clarification messages. For instance: `"getDistinctRowCount(, groupKey=" + col1 + ", null) should match ColStatistics"`.

File: core/src/test/java/org/apache/calcite/test/RelMetadataTest.java [hunk 8]
```diff
@@ -3102,6 +3108,69 @@ private void checkNodeTypeCount(String sql, Map<Class<? extends RelNode>, Intege
...
+
+    Double size = mq.getAverageRowSize(rel);
+    assertThat(size.intValue(), is(6));
```
Comment (vlsi) [inline core/src/test/java/org/apache/calcite/test/RelMetadataTest.java:3171]: Please add the clarification messages. For instance: `"getAverageRowSize(...), rowType=" + rel.getRowType()`

File: core/src/main/java/org/apache/calcite/rel/metadata/RelMdDistinctRowCount.java [hunk 9]
```diff
@@ -79,6 +82,37 @@ public Double getDistinctRowCount(RelNode rel, RelMetadataQuery mq,
     return null;
   }
 
+  public Double getDistinctRowCount(TableScan rel, RelMetadataQuery mq,
+      ImmutableBitSet groupKey, RexNode predicate) {
+    final double selectivity = predicate == null ? 1D : mq.getSelectivity(rel, predicate);
```
Comment (vlsi) [inline core/src/main/java/org/apache/calcite/rel/metadata/RelMdDistinctRowCount.java:87]: Can `getSelectivity` return null?

**Comment by chunweilei** [comment]
It's outdated. See https://github.com/apache/calcite/pull/2845.

**Pull request closed without merging**